Issue with Neo4j kafka sink connector in combination with Avro schema nesting

(edited, added full stacktrace)

Hi,

I'm trying to ingest Kafka messages encoded in Avro into neo4j using the kafka connect neo4j sink.

I'm getting this error:
 
> Caused by: org.apache.kafka.connect.errors.DataException: org.neo4j.driver.internal.value.FloatValue is not a supported data type.

This error seems to be caused by a "double" datatype in my Avro schema:

> Caused by: org.apache.kafka.connect.errors.DataException: Exception thrown while processing field 'latitude'

Stacktrace:

```
[2021-11-09 16:52:53,416] ERROR ErrorData(originalTopic=rtim-wagon-gps, timestamp=1603019642492, partition=8, offset=295990, exception=org.apache.kafka.connect.errors.DataException: Exception thrown while processing field 'networkLocation', key=338849101419|3187547f-4a5a-4f02-acf0-7bd617712fd6, value=Struct{parentMessageUuid=f0a2724f-656b-43fd-a1fa-22273e4335d7,wagonNumber=338849101419,measurementSource=WAGON_GPS,measurementTime=2020-10-18T13:10:00+02:00,measurementType=WAGON_GPS,messageTime=2020-, executingClass=class streams.kafka.connect.sink.Neo4jSinkTask) (streams.kafka.connect.sink.Neo4jSinkTask)
org.apache.kafka.connect.errors.DataException: Exception thrown while processing field 'networkLocation'
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertStruct(AbstractConverter.java:259)
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convert(AbstractConverter.java:78)
	at streams.kafka.connect.utils.ConnectExtensionFunctionsKt.convertData(ConnectExtensionFunctions.kt:24)
	at streams.kafka.connect.utils.ConnectExtensionFunctionsKt.convertData$default(ConnectExtensionFunctions.kt:23)
	at streams.kafka.connect.utils.ConnectExtensionFunctionsKt.toStreamsSinkEntity(ConnectExtensionFunctions.kt:19)
	at streams.kafka.connect.sink.EventBuilder.build(EventBuilder.kt:28)
	at streams.kafka.connect.sink.Neo4jSinkTask.put(Neo4jSinkTask.kt:41)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:560)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:323)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:226)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:198)
	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:185)
	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:235)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.apache.kafka.connect.errors.DataException: Exception thrown while processing field 'coordinates'
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertStruct(AbstractConverter.java:259)
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convert(AbstractConverter.java:78)
	at streams.kafka.connect.sink.converters.MapValueConverter.setStructField(MapValueConverter.kt:56)
	at streams.kafka.connect.sink.converters.MapValueConverter.setStructField(MapValueConverter.kt:9)
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertStruct(AbstractConverter.java:244)
	... 17 more
Caused by: org.apache.kafka.connect.errors.DataException: Exception thrown while processing field 'latitude'
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertMap(AbstractConverter.java:159)
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convert(AbstractConverter.java:80)
	at streams.kafka.connect.sink.converters.MapValueConverter.setMap(MapValueConverter.kt:62)
	at streams.kafka.connect.sink.converters.MapValueConverter.setStructField(MapValueConverter.kt:57)
	at streams.kafka.connect.sink.converters.MapValueConverter.setStructField(MapValueConverter.kt:9)
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertStruct(AbstractConverter.java:244)
	... 21 more
Caused by: org.apache.kafka.connect.errors.DataException: org.neo4j.driver.internal.value.FloatValue is not a supported data type.
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertMap(AbstractConverter.java:151)
	... 26 more
```

I believe that the schema is parsed incorrectly, where a Double type is converted into a org.neo4j.driver.internal.value.FloatValue, although the AbstractConverter class is not able to handle this datatype, but can only process java.lang.Float.

I'm deploying containers
- version 2.0.0 of the neo4j sink connector
- version 6.0.0 of the community confluent platform
- neo4j 4.1.9 (latest available docker image)

This is the Avro schema of records on the Kafka cluster, where the latitude field is configured as a double

```
{
  "type": "record",
  "name": "WagonGpsMeasurement",
  "namespace": "be.lineas.rtim.model.gps",
  "fields": [
    {
      "name": "parentMessageUuid",
      "type": [
        "null",
        {
          "type": "string",
          "CustomEncoding": "NullableUUIDEncoding"
        }
      ],
      "default": null
    },
    {
      "name": "positionAccuracy",
      "type": [
        "null",
        "long"
      ],
      "default": null
    },
    {
      "name": "positionProvider",
      "type": [
        "null",
        "string"
      ],
      "default": null
    },
    {
      "name": "assignedWagonTripKey",
      "type": [
        "null",
        {
          "type": "record",
          "name": "WagonTripKey",
          "namespace": "be.lineas.rtim.model.wagonevent",
          "fields": [
            {
              "name": "start",
              "type": {
                "type": "string",
                "CustomEncoding": "InstantEncoding"
              }
            },
            {
              "name": "wagon",
              "type": "string"
            }
          ]
        }
      ],
      "default": null
    },
    {
      "name": "wagonNumber",
      "type": "string"
    },
    {
      "name": "assignedTrainId",
      "type": [
        "null",
        {
          "type": "record",
          "name": "TrainId",
          "namespace": "be.lineas.rtim.model.train",
          "fields": [
            {
              "name": "internationalTrainNumber",
              "type": "string"
            },
            {
              "name": "trainDate",
              "type": {
                "type": "string",
                "CustomEncoding": "LocalDateEncoding"
              }
            }
          ]
        }
      ],
      "default": null
    },
    {
      "name": "legUuid",
      "type": [
        "null",
        {
          "type": "string",
          "CustomEncoding": "NullableUUIDEncoding"
        }
      ],
      "default": null
    },
    {
      "name": "measurementSource",
      "type": {
        "type": "enum",
        "name": "MeasurementSource",
        "namespace": "be.lineas.rtim.model.measurement",
        "symbols": [
          "RTIS",
          "WAGON_GPS",
          "ISR"
        ]
      }
    },
    {
      "name": "measurementTime",
      "type": [
        "null",
        {
          "type": "string",
          "CustomEncoding": "NullableOffsetDateTimeEncoding"
        }
      ],
      "default": null
    },
    {
      "name": "measurementType",
      "type": {
        "type": "enum",
        "name": "MeasurementType",
        "namespace": "be.lineas.rtim.model.measurement",
        "symbols": [
          "DEPARTURE",
          "PASSAGE",
          "ARRIVAL",
          "WAGON_GPS",
          "LEFT_ORIGIN",
          "WAGON_ARRIVAL",
          "WAGON_DEPARTURE",
          "REACHED_DESTINATION",
          "PASSED_TROUGH",
          "BORDER_CROSSING",
          "BORDER_CROSSING_EXIT",
          "BORDER_CROSSING_ENTRY",
          "BORDER_CROSSING_ADVICE",
          "WAGON_REFUSED",
          "READY_TO_PULL",
          "PULLED",
          "DELIVERED",
          "FREE_FOR_DISPOSITION",
          "WAGON_BROKEN",
          "WAGON_REPAIRED",
          "TRANSPORT_CREATED",
          "CREATE_TRANSPORT_DOSSIER",
          "UPDATE_TRANSPORT_DOSSIER",
          "WAGON_MILEAGE_CALCULATION_RESULT",
          "DELETE_TRANSPORT_DOSSIER",
          "ELTA_FORWARDING"
        ]
      }
    },
    {
      "name": "messageTime",
      "type": [
        "null",
        {
          "type": "string",
          "CustomEncoding": "NullableOffsetDateTimeEncoding"
        }
      ],
      "default": null
    },
    {
      "name": "messageUuids",
      "type": {
        "type": "array",
        "items": "string",
        "CustomEncoding": "UUIDListEncoding"
      }
    },
    {
      "name": "networkLocation",
      "type": [
        "null",
        {
          "type": "record",
          "name": "NetworkLocation",
          "namespace": "be.lineas.rtim.model.networklocation",
          "fields": [
            {
              "name": "coordinates",
              "type": [
                "null",
                {
                  "type": "record",
                  "name": "Coordinates",
                  "namespace": "be.lineas.rtim.model.gps",
                  "fields": [
                    {
                      "name": "latitude",
                      "type": "double"
                    },
                    {
                      "name": "longitude",
                      "type": "double"
                    }
                  ]
                }
              ],
              "default": null
            },
            {
              "name": "countryIsoCode",
              "type": [
                "null",
                "string"
              ],
              "default": null
            },
            {
              "name": "isEnriched",
              "type": "boolean"
            },
            {
              "name": "locationUUID",
              "type": [
                "null",
                {
                  "type": "string",
                  "CustomEncoding": "NullableUUIDEncoding"
                }
              ],
              "default": null
            },
            {
              "name": "name",
              "type": [
                "null",
                "string"
              ],
              "default": null
            },
            {
              "name": "netCode",
              "type": [
                "null",
                "string"
              ],
              "default": null
            },
            {
              "name": "primaryLocationCode",
              "type": [
                "null",
                "string"
              ],
              "default": null
            },
            {
              "name": "uicCode",
              "type": [
                "null",
                "string"
              ],
              "default": null
            }
          ]
        }
      ],
      "default": null
    },
    {
      "name": "uuid",
      "type": {
        "type": "string",
        "CustomEncoding": "UUIDEncoding"
      }
    }
  ]
}
```

Hi @mroiter-larus,

I've tested with another message and schema, I get a similar error but it seems not to be related to doubles ...

The schema:

```
{
  "type": "record",
  "name": "record",
  "fields": [
    {
      "name": "trainId",
      "type": {
        "type": "record",
        "name": "TrainId",
        "namespace": "be.lineas.rtim.model.train",
        "fields": [
          {
            "name": "internationalTrainNumber",
            "type": "string"
          },
          {
            "name": "trainDate",
            "type": "string"
          }
        ]
      }
    },
    {
      "name": "coreId",
      "type": [
        "null",
        "string"
      ]
    }
  ]
}
```
Example message:
```
{
      "trainId": {
        "internationalTrainNumber": "46261",
        "trainDate": "2021-05-20"
      },
      "coreId": "000000046261"
    }
```
Stacktrace:
```
org.apache.kafka.connect.errors.DataException: Exception thrown while processing field 'trainId'
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertStruct(AbstractConverter.java:259)
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convert(AbstractConverter.java:78)
	at streams.kafka.connect.utils.ConnectExtensionFunctionsKt.convertData(ConnectExtensionFunctions.kt:24)
	at streams.kafka.connect.utils.ConnectExtensionFunctionsKt.convertData$default(ConnectExtensionFunctions.kt:23)
	at streams.kafka.connect.utils.ConnectExtensionFunctionsKt.toStreamsSinkEntity(ConnectExtensionFunctions.kt:19)
	at streams.kafka.connect.sink.EventBuilder.build(EventBuilder.kt:28)
	at streams.kafka.connect.sink.Neo4jSinkTask.put(Neo4jSinkTask.kt:41)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:582)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:330)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:232)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:201)
	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:188)
	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:237)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.kafka.connect.errors.DataException: Exception thrown while processing field 'internationalTrainNumber'
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertMap(AbstractConverter.java:159)
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convert(AbstractConverter.java:80)
	at streams.kafka.connect.sink.converters.MapValueConverter.setMap(MapValueConverter.kt:62)
	at streams.kafka.connect.sink.converters.MapValueConverter.setStructField(MapValueConverter.kt:57)
	at streams.kafka.connect.sink.converters.MapValueConverter.setStructField(MapValueConverter.kt:9)
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertStruct(AbstractConverter.java:244)
	... 17 more
Caused by: org.apache.kafka.connect.errors.DataException: org.neo4j.driver.internal.value.StringValue is not a supported data type.
	at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertMap(AbstractConverter.java:151)
	... 22 more
```

This error occurs irrespective of what cypher query I configure in the connector. Seems to me like this is related to the fact that the "internationalTrainNumber" field is nested in the "trainId" field? I was trying to set up a unit test using a nested AVRO message on the streams project to check this, but the code structure is not that trivial. I'll keep investigating though, but any help is appreciated.

Cheers,
Thomas

--- a/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
+++ b/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
@@ -1,5 +1,6 @@
 package streams.kafka.connect.sink.converters
 
+import org.apache.kafka.connect.data.Struct
 import org.neo4j.driver.Value
 import org.neo4j.driver.Values
 import java.math.BigDecimal
@@ -51,5 +52,12 @@ class Neo4jValueConverter: MapValueConverter<Value>() {
     override fun setDateField(result: MutableMap<String, Value?>?, fieldName: String, value: Date) {
         val localDate = value.toInstant().atZone(UTC).toLocalDate()
         setValue(result, fieldName, localDate)
+    }
+
+    override fun setStructField(result: MutableMap<String, Value?>?, fieldName: String, value: Struct) {
+        val converted = convert(value)
+            .mapValues { it.value?.asObject() }
+            .toMutableMap() as MutableMap<Any?, Any?>
+        setMap(result, fieldName, null, converted)
     }
 }

--- a/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
+++ b/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
@@ -181,6 +181,25 @@ class Neo4jValueConverterTest {
         assertEquals(date.toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime(), result["date"]?.asLocalDateTime())
     }
 
+    @Test
+    fun `should be able to process a nested AVRO structure`() {
+        val trainSchema = SchemaBuilder.struct()
+            .field("internationalTrainNumber", Schema.STRING_SCHEMA)
+            .field("trainDate", Schema.STRING_SCHEMA).build()
+        val mySchema = SchemaBuilder.struct()
+            .field("trainId", trainSchema)
+            .field("coreId", Schema.STRING_SCHEMA).build()
+
+        val trainIdStruct = Struct(trainSchema)
+            .put("internationalTrainNumber", "46261")
+            .put("trainDate", "2021-05-20")
+        val rootStruct = Struct(mySchema)
+            .put("trainId", trainIdStruct)
+            .put("coreId", "000000046261")
+
+        val result = Neo4jValueConverter().convert(rootStruct) as Map<*, *>
+    }
+
     companion object {
         private val LI_SCHEMA = SchemaBuilder.struct().name("org.neo4j.example.html.LI")
                 .field("value", Schema.OPTIONAL_STRING_SCHEMA)

